lib/streamlit/runtime/connection_factory.py

```diff
@@ -0,0 +1,209 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import importlib
+import os
+import re
+from datetime import timedelta
+from typing import Any, Dict, Type, TypeVar, overload
+
+from typing_extensions import Final, Literal
+
+from streamlit.connections import SQL, ExperimentalBaseConnection, Snowpark
+from streamlit.errors import StreamlitAPIException
+from streamlit.runtime.caching import cache_resource
+from streamlit.runtime.metrics_util import gather_metrics
+from streamlit.runtime.secrets import secrets_singleton
+
+# NOTE: Adding support for a new first party connection requires:
+#   1. Adding the new connection name and class to this dict.
+#   2. Writing a new @overload for connection_factory.
+#   3. Updating test_get_first_party_connection_helper in connection_factory_test.py.
+FIRST_PARTY_CONNECTIONS = {
+    "snowpark": Snowpark,
+    "sql": SQL,
+}
+MODULE_EXTRACTION_REGEX = re.compile(r"No module named \'(.+)\'")
+MODULES_TO_PYPI_PACKAGES: Final[Dict[str, str]] = {
+    "sqlalchemy": "sqlalchemy",
+    "snowflake": "snowflake-snowpark-python",
+    "snowflake.snowpark": "snowflake-snowpark-python",
+    "tenacity": "tenacity",
```

Comment:
Does it mean the developer will often need to manually install tenacity to get the SQL or Snowpark connection working?

I think it's an OK tradeoff to avoid a hard tenacity dependency on core, but it'll be a pretty annoying experience since this isn't at all obvious to the developer.

Curious if anyone else has bright ideas to address this. I don't think it's ship blocking but should definitely be addressed in docs and maybe consider a fast follow.

Reply:
Had a moment of concern but thankfully looks like `tenacity` is already in the [snowflake conda channel](https://repo.anaconda.com/pkgs/snowflake/), so hopefully no concerns or issues about using it there (and also might be a vote of confidence for just adding it as a core dependency 😉)

![image](https://user-images.githubusercontent.com/116604821/232171740-b2362fb8-b636-46b9-bd4b-ff9e7c1cca63.png)